pkg/export/secrets/manager.go

```diff
@@ -0,0 +1,233 @@
+	// If no secrets are provided, cancel any existing secret provider.
+	if len(configs) == 0 {
+		m.cancelFn()
+		m.provider = nil
+		m.cancelFn = func() {}
+		m.secrets = map[string]*secretEntry{}
+		m.config = nil
+		return nil
+	}
+
+	eq, err := yamlEqual(m.config, providerConfig)
+	if err != nil {
+		return err
+	}
+
+	defer func() {
+		m.config = providerConfig
+	}()
+
+	if !eq || m.provider == nil {
+		ctx, cancel := context.WithCancel(m.ctx)
+		provider, err := providerConfig.newProvider(ctx, m.opts)
+		if err != nil {
+			cancel()
+			return err
+		}
+
+		m.cancelFn()
+		m.provider = provider
+		m.cancelFn = cancel
+		m.secrets = map[string]*secretEntry{}
+	}
+	return m.updateSecrets(ctx, configs)
+}
+
+func (m *Manager) updateSecrets(ctx context.Context, configs []SecretConfig) error {
+	var errs []error
+
+	// Do a first pass to check for errors and disable those secrets.
+	secretNamesEnabled := make(map[string]bool)
+	for _, secret := range configs {
+		if enabled, ok := secretNamesEnabled[secret.Name]; ok {
+			if !enabled {
+				continue
+			}
+			errs = append(errs, fmt.Errorf("duplicate secret key %q", secret.Name))
+			secretNamesEnabled[secret.Name] = false
+		} else {
+			secretNamesEnabled[secret.Name] = true
+		}
+	}
+
+	secretsFinal := map[string]*secretEntry{}
+	for i := range configs {
+		secretIncoming := &configs[i]
+		if enabled := secretNamesEnabled[secretIncoming.Name]; !enabled {
+			continue
+		}
+		// First check if we've registered this secret before.
+		if secretPrevious, ok := m.secrets[secretIncoming.Name]; ok {
+			// Track all the secrets we saw. The leftover are later removed.
+			delete(m.secrets, secretIncoming.Name)
+
+			// If the config didn't change, we skip this one.
+			eq, err := yamlEqual(&secretPrevious.config, &secretIncoming.Config)
+			if err != nil {
+				errs = append(errs, err)
+				continue
+			}
+			if eq {
+				secretsFinal[secretIncoming.Name] = secretPrevious
+				continue
+			}
+
+			// The config changed, so update it.
+			s, err := m.provider.Update(ctx, &secretPrevious.config, &secretIncoming.Config)
+			if err != nil {
+				errs = append(errs, err)
+				continue
+			}
+			secretPrevious.secret = s
+			secretsFinal[secretIncoming.Name] = secretPrevious
+		} else {
+			// We've never seen this secret before, so add it.
+			s, err := m.provider.Add(ctx, &secretIncoming.Config)
+			if err != nil {
+				errs = append(errs, err)
+				continue
+			}
+			secretsFinal[secretIncoming.Name] = &secretEntry{
+				config: secretIncoming.Config,
+				secret: s,
+			}
+		}
+	}
+	for _, secretUnused := range m.secrets {
+		if err := m.provider.Remove(&secretUnused.config); err != nil {
+			errs = append(errs, err)
+		}
+	}
+
+	m.secrets = secretsFinal
+
+	total := len(secretNamesEnabled)
+	success := len(m.secrets)
+	failedSecretConfigs.Set(float64(total - success))
+	secretsTotal.Set(float64(total))
+	return errors.Join(errs...)
+}
+
+// Fetch implements Secret.Fetch.
```

Comment:
question: Is this true? `Secret.Fetch` just takes in a `context.Context` argument.

Reply:
You are right, it actually implements `github.com/prometheus/common/config.SecretManager.Fetch`.